File: fluid.py
import time
import random
import sys

# Simulation constants
SIM_WIDTH = 128
SIM_HEIGHT = 64
NUM_PARTICLES = 200
GRAVITY = 0.5
DAMPING = 0.82
BORDER = 2.0
FRAME_DELAY = 0.033  # ~30 FPS

class Particle:
    def __init__(self):
        self.x = random.uniform(BORDER, SIM_WIDTH - BORDER)
        self.y = random.uniform(BORDER, SIM_HEIGHT/2 - BORDER)
        self.vx = random.uniform(-1, 1) * 0.7
        self.vy = random.uniform(-0.5, 0.5)

def initialize_particles(num_particles):
    return [Particle() for _ in range(num_particles)]

def update_physics(particles):
    for p in particles:
        # Apply gravity
        p.vy += GRAVITY
        
        # Update position
        p.x += p.vx
        p.y += p.vy

        # Horizontal boundary collisions
        if p.x < BORDER or p.x >= SIM_WIDTH - BORDER:
            p.vx *= -DAMPING
            p.x = max(BORDER, min(p.x, SIM_WIDTH - BORDER - 0.1))

        # Vertical boundary collisions
        if p.y < BORDER or p.y >= SIM_HEIGHT - BORDER:
            p.vy *= -DAMPING
            p.y = max(BORDER, min(p.y, SIM_HEIGHT - BORDER - 0.1))

def draw_frame(particles):
    # Initialize empty grid
    grid = [[' ' for _ in range(SIM_WIDTH)] for _ in range(SIM_HEIGHT)]
    
    # Plot particles
    for p in particles:
        x = int(p.x)
        y = int(p.y)
        if 0 <= x < SIM_WIDTH and 0 <= y < SIM_HEIGHT:
            grid[y][x] = 'o'
    
    # Build frame buffer
    buffer = []
    for row in grid:
        buffer.append(''.join(row))
    
    # Clear screen and move cursor to top-left
    sys.stdout.write('\033[H\033[J')
    sys.stdout.write('\n'.join(buffer))
    sys.stdout.flush()

def main():
    particles = initialize_particles(NUM_PARTICLES)
    
    # Hide cursor
    sys.stdout.write('\033[?25l')
    sys.stdout.flush()

    try:
        while True:
            start_time = time.monotonic()
            
            update_physics(particles)
            draw_frame(particles)
            
            # Frame rate control
            elapsed = time.monotonic() - start_time
            sleep_time = FRAME_DELAY - elapsed
            if sleep_time > 0:
                time.sleep(sleep_time)
                
    except KeyboardInterrupt:
        pass
    finally:
        # Show cursor before exiting
        sys.stdout.write('\033[?25h')
        sys.stdout.flush()

if __name__ == "__main__":
    main()

File: fluid.cpp
#include <iostream>
#include <vector>
#include <chrono>
#include <cstdlib>
#include <cmath>

using namespace std;

// Simulation constants
const int SIM_WIDTH = 64;
const int SIM_HEIGHT = 32;
const int NUM_PARTICLES = 100;

struct Particle {
    double x, y;
    double vx, vy;
};

vector<Particle> particles(NUM_PARTICLES);

void initializeParticles() {
    for (auto& p : particles) {
        p.x = rand() % SIM_WIDTH;
        p.y = rand() % (SIM_HEIGHT / 2);
        p.vx = (rand() % 200 - 100) / 100.0 * 0.7;
        p.vy = (rand() % 100 - 50) / 100.0;
    }
}

void updatePhysics() {
    const double gravity = 0.15;
    const double damping = 0.82;
    const double border = 2.0;

    for (auto& p : particles) {
        p.vy += gravity;
        p.x += p.vx;
        p.y += p.vy;

        // Horizontal boundary collisions
        if (p.x < border || p.x >= SIM_WIDTH - border) {
            p.vx = -p.vx * damping;
            p.x = max(border, min(p.x, SIM_WIDTH - border - 0.1));
        }

        // Vertical boundary collisions
        if (p.y < border || p.y >= SIM_HEIGHT - border) {
            p.vy = -p.vy * damping;
            p.y = max(border, min(p.y, SIM_HEIGHT - border - 0.1));
        }
    }
}

void drawFrame() {
    vector<vector<char> > grid(SIM_HEIGHT, vector<char>(SIM_WIDTH, ' '));

    // Plot particles
    for (const auto& p : particles) {
        int x = static_cast<int>(p.x);
        int y = static_cast<int>(p.y);
        x = max(0, min(SIM_WIDTH - 1, x));
        y = max(0, min(SIM_HEIGHT - 1, y));
        grid[y][x] = 'o';
    }

    // Clear screen and reset cursor
    cout << "\033[H";

    // Draw grid
    for (const auto& row : grid) {
        for (char c : row) {
            cout << c;
        }
        cout << '\n';
    }
    cout << flush;
}

int main() {
    srand(time(nullptr));
    initializeParticles();
    
    // Hide cursor
    cout << "\033[?25l";

    auto last_frame = chrono::steady_clock::now();
    const chrono::milliseconds frame_delay(33);

    try {
        while (true) {
            auto now = chrono::steady_clock::now();
            if (now - last_frame >= frame_delay) {
                updatePhysics();
                drawFrame();
                last_frame = now;
            }
        }
    } catch (...) {
        // Show cursor before exiting
        cout << "\033[?25h";
    }

    // Restore cursor
    cout << "\033[?25h";
    return 0;
}
